src/Agent.Worker/Handlers/ProcessHandler.cs

```diff
@@ -132,32 +132,69 @@ public async Task RunAsync()
 
-            var enableFileArgs = disableInlineExecution && enableSecureArguments;
-
-            if ((disableInlineExecution && (enableSecureArgumentsAudit || enableSecureArguments)) || enableTelemetry)
+            bool enableFileArgs = disableInlineExecution && enableSecureArguments && !enableNewPHLogic;
+            if (enableFileArgs)
             {
-                var (processedArgs, telemetry) = ProcessHandlerHelper.ProcessInputArguments(arguments);
+                bool enableSecureArgumentsAudit = AgentKnobs.ProcessHandlerSecureArgumentsAudit.GetValue(ExecutionContext).AsBoolean();
+                ExecutionContext.Debug($"Enable secure arguments audit: '{enableSecureArgumentsAudit}'");
+                bool enableTelemetry = AgentKnobs.ProcessHandlerTelemetry.GetValue(ExecutionContext).AsBoolean();
+                ExecutionContext.Debug($"Enable telemetry: '{enableTelemetry}'");
+
+                if ((disableInlineExecution && (enableSecureArgumentsAudit || enableSecureArguments)) || enableTelemetry)
+                {
+                    var (processedArgs, telemetry) = ProcessHandlerHelper.ExpandCmdEnv(arguments, Environment);
 
-                if (disableInlineExecution && enableSecureArgumentsAudit)
+                    if (disableInlineExecution && enableSecureArgumentsAudit)
+                    {
+                        ExecutionContext.Warning($"The following arguments will be executed: '{processedArgs}'");
+                    }
+                    if (enableFileArgs)
+                    {
+                        GenerateScriptFile(cmdExe, command, processedArgs);
+                    }
+                    if (enableTelemetry)
+                    {
+                        ExecutionContext.Debug($"Agent PH telemetry: {JsonConvert.SerializeObject(telemetry.ToDictionary(), Formatting.None)}");
+                        PublishTelemetry(telemetry.ToDictionary(), "ProcessHandler");
+                    }
+                }
+            }
+            else if (enableNewPHLogic)
+            {
+                try
                 {
-                    ExecutionContext.Warning($"The following arguments will be executed: '{processedArgs}'");
+                    var (isValid, telemetry) = ProcessHandlerHelper.ValidateInputArguments(arguments, Environment, ExecutionContext);
+
+                    if (telemetry != null)
+                    {
+                        PublishTelemetry(telemetry, "ProcessHandler");
+                    }
+                    if (!isValid)
+                    {
+                        throw new ArgsSanitizedException(StringUtil.Loc("ProcessHandlerScriptArgsSanitized"));
+                    }
                 }
-                if (enableFileArgs)
+                catch (ArgsSanitizedException)
                 {
-                    GenerateScriptFile(cmdExe, command, processedArgs);
+                    throw;
                 }
```

Comment:
not blocking: this could move this out of the try block; then  the extra catch block isn't needed

Reply:
updated. now it throws by `shouldThrow`